hashing/te-tag-query-java/com/facebook/threatexchange/DescriptorPostParameters.java

```diff
@@ -330,60 +330,58 @@ public String getReactionsToRemove() {
     return this._reactionsToRemove;
   }
 
-  public boolean validateForSubmitWithReport(PrintStream o) {
+  // Returns null if no errors, non-null for error.
+  public String validateForSubmitWithReport() {
     if (this._descriptorID != null) {
-      o.println("Descriptor ID must not be specified for submit.\n");
-      return false;
+      return "Descriptor ID must not be specified for submit.";
     }
     if (this._indicatorText == null) {
-      o.println("Indicator text is missing.\n");
-      return false;
+      return "Indicator text is missing.";
     }
     if (this._indicatorType == null) {
-      o.println("Indicator type is missing.\n");
-      return false;
+      return "Indicator type is missing.";
     }
     if (this._description == null) {
-      o.println("Description is missing.\n");
-      return false;
+      return "Description is missing.";
     }
     if (this._shareLevel == null) {
-      o.println("Share level is missing.\n");
-      return false;
+      return "Share level is missing.\n";
     }
     if (this._status == null) {
-      o.println("Status is missing.\n");
-      return false;
+      return "Status is missing.\n";
     }
     if (this._privacyType == null) {
-      o.println("Privacy type is missing.\n");
-      return false;
+      return "Privacy type is missing.\n";
     }
-    return true;
+    return null;
   }
 
-  public boolean validateForUpdateWithReport(PrintStream o) {
+  // Returns null if no errors, non-null for error.
+  public String validateForUpdateWithReport() {
     if (this._descriptorID == null) {
-      o.println("Descriptor ID must be specified for update.\n");
-      return false;
+      return "Descriptor ID must be specified for update.";
     }
     if (this._indicatorText != null) {
-      o.println("Indicator text must not be specified for update.\n");
-      return false;
+      return "Indicator text must not be specified for update.";
     }
     if (this._indicatorType != null) {
-      o.println("Indicator type must not be specified for update.\n");
-      return false;
+      return "Indicator type must not be specified for update.";
     }
-    return true;
+    return null;
   }
 
-  public boolean validateForCopyWithReport(PrintStream o) {
+  // Returns null if no errors, non-null for error.
+  public String validateForCopyWithReport() {
     if (this._descriptorID == null) {
-      o.println("Descriptor ID must be specified for copy.\n");
-      return false;
+      return "Descriptor ID must be specified for copy.";
     }
-    return true;
+    if (this._privacyMembers == null) {
```

Comment:
are privacyMembers required for copy but not submit?

Reply:
There's a TE API constraint at present that only the owner of a descriptor can read back the privacy members over the API. That's the only reason you need to specify it explictly, since if you don't own the copy-from descriptor, you can't copy its visibility. I'll add a comment to clarify.